### ISSUE TITLE
the output image of test.py and the intermediate image saved during training are different?

### ISSUE THREAD

**Issue by JianqiangRen** [NVlabs/SPADE#37]
![aa](https://user-images.githubusercontent.com/12150021/56726916-473b7880-6782-11e9-84cd-bed53891863a.png)
I trained  with some landscape images and label maps(use_vae:  False). During training, I find intermediate images in checkpoint/web is great(third column), but when I test on the same label map using test.py, the output are different(last column). Is that because when saving intermediate results during training, VAE is always enabled regardless of whether use_vae option are set? Could you explain it? 
Thank!

**Comment by taesungp** [comment]
VAE should not be enabled if it was not specified in the option. In fact, the generator architecture is different depending on whether VAE is enabled or not, so it should be impossible to switch from one setting to another between training and test time. 

It might be related to the discrepancy in batch norm between training and test time. At training time, it uses the batch statistics of the minibatch. At test time, it uses the saved batch statistics from checkpoint. This feature can be turned on and off by commenting `.eval()` in https://github.com/NVlabs/SPADE/blob/master/test.py#L20. 

At the same time, I think your model has overfitted to the dataset. What's the size of the dataset?

**Comment by JianqiangRen** [comment]
> VAE should not be enabled if it was not specified in the option. In fact, the generator architecture is different depending on whether VAE is enabled or not, so it should be impossible to switch from one setting to another between training and test time.
> 
> It might be related to the discrepancy in batch norm between training and test time. At training time, it uses the batch statistics of the minibatch. At test time, it uses the saved batch statistics from checkpoint. This feature can be turned on and off by commenting `.eval()` in https://github.com/NVlabs/SPADE/blob/master/test.py#L20.
> 
> At the same time, I think your model has overfitted to the dataset. What's the size of the dataset?

Yes! After commented  .eval(), the result become more vivid like origin real image. Below illustrates the comparison of  turning on/off using saved batch statistics. As we can see, when using saved batch statistic, the results are dull, such as the color of sky and grass are always the same among all the outputs.
![屏幕快照 2019-04-28 下午4 50 00](https://user-images.githubusercontent.com/12150021/56861675-af35cd00-69d5-11e9-8340-def0f1bbcbfb.png)
Since the batch size for training and test are 1, should we comment .eval() when testing? what's GauGAN setting? 
My training dataset contains 1400 images, I think my model does overfitted to it, but without  sufficient training epochs, the result always have a lot of artifacts. Maybe I need more images?

**Comment by taesungp** [comment]
For GauGAN, we trained with batch size 32, and at test time we turned `eval()` on with batch size 1. 

Note that commenting out `eval()` will also affect spectral normalization. In detail, it will actually change the weights of the network every time you call forward. This may result in drifting in output image quality as you call many forward passes.
